compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5390,6 +5395,28 @@ private BType checkInvocationArgs(BLangInvocation iExpr, List<BType> paramTypes,
         }
     }
 
+    private void checkArrayLibSortFuncArgs(BLangInvocation iExpr) {
+        if (iExpr.argExprs.size() <= 2) {
+            if (!types.isBasicType(iExpr.expr.type)) {
+                dlog.error(iExpr.expr.pos, DiagnosticCode.INVALID_SORT_ARRAY_MEMBER_TYPE, iExpr.expr.type);
+            }
+        } else if (iExpr.argExprs.size() == 3) {
+            if (iExpr.argExprs.get(2).type != null) {
+                if (iExpr.argExprs.get(2).type.tag == TypeTags.SEMANTIC_ERROR) {
+                    return;
+                } else if (iExpr.argExprs.get(2).type.tag == TypeTags.NIL && !types.isBasicType(iExpr.expr.type)) {
+                    dlog.error(iExpr.expr.pos, DiagnosticCode.INVALID_SORT_ARRAY_MEMBER_TYPE, iExpr.expr.type);
+                }
+            } else {
+                BLangLambdaFunction func = (BLangLambdaFunction) iExpr.argExprs.get(2);
+                BType returnType = func.function.type.getReturnType();
+                if (!types.isBasicType(returnType)) {
+                    dlog.error(iExpr.expr.pos, DiagnosticCode.INVALID_SORT_FUNC_RETURN_TYPE, returnType);
+                }
+            }
+        }
+    }
```

Comment:
I think we can refactor this method a bit by returning early on error and merging the if blocks.
```suggestion
    private void checkArrayLibSortFuncArgs(BLangInvocation iExpr) {
        if (iExpr.argExprs.size() <= 2 && !types.isBasicType(iExpr.expr.type)) {
            dlog.error(iExpr.expr.pos, DiagnosticCode.INVALID_SORT_ARRAY_MEMBER_TYPE, iExpr.expr.type);
        } 
        
        if (iExpr.argExprs.size() != 3) {
            return;
        }

        BLangExpression argThree = iExpr.argExprs.get(2);
        BType argThreeType = argThree.type;
        if (argThreeType != null) {
            if (argThreeType.tag == TypeTags.SEMANTIC_ERROR) {
                return;
            }
            
            if (argThreeType.tag == TypeTags.NIL && !types.isBasicType(iExpr.expr.type)) {
                dlog.error(iExpr.expr.pos, DiagnosticCode.INVALID_SORT_ARRAY_MEMBER_TYPE, iExpr.expr.type);
                return;
            }
        }
        
        BLangLambdaFunction func = (BLangLambdaFunction) argThree;
        BType returnType = func.function.type.getReturnType();
        if (!types.isBasicType(returnType)) {
            dlog.error(iExpr.expr.pos, DiagnosticCode.INVALID_SORT_FUNC_RETURN_TYPE, returnType);
        }
    }
```